src/query/storage/m3/multi_fetch_result.go

```diff
@@ -63,23 +70,44 @@ type multiResultSeries struct {
 }
 
 func (r *multiResult) Close() error {
+	r.Lock()
+	defer r.Unlock()
+
 	for _, iters := range r.seenIters {
 		iters.Close()
 	}
+	r.seenIters = nil
+
+	if r.finalResult != nil {
+		// NB(r): Since all the series iterators in the final result are held onto
```

Comment:
Mind adding a unit test w/ mocks to ensure we free the underlying iterators exactly once assuming the API is used correctly. I think it's 100% right the way it is but the test will be good for when someone comes and attempts to refactors this

Reply:
As discussed, let's revisit this in a followup change.